File: LLM-linguistic-translations/OpenAi-chatgpt/o4/py-translations/chat-code.py
from typing import Union, Dict, Callable, List, Tuple as TypingTuple, Optional, Any
import ipaddress
import csv
import sys
from collections import defaultdict
from dataclasses import dataclass

# --- op_result variants ---
class OpResult: pass

@dataclass(frozen=True)
class FloatResult(OpResult):
    value: float

@dataclass(frozen=True)
class IntResult(OpResult):
    value: int

@dataclass(frozen=True)
class IPv4Result(OpResult):
    value: ipaddress.IPv4Address

@dataclass(frozen=True)
class MACResult(OpResult):
    value: bytes

class EmptyResult(OpResult): pass

TupleType = Dict[str, OpResult]

# --- Operator ---
class Operator:
    def __init__(self, next_fn: Callable[[TupleType], None], reset_fn: Callable[[TupleType], None]) -> None:
        self.next = next_fn
        self.reset = reset_fn

# --- Conversions ---
def string_of_mac(buf: bytes) -> str:
    return ":".join(f"{b:02x}" for b in buf)

TCP_FLAGS = [
    ("FIN", 1<<0), ("SYN", 1<<1), ("RST", 1<<2), ("PSH", 1<<3),
    ("ACK",1<<4), ("URG",1<<5), ("ECE",1<<6), ("CWR",1<<7)
]

def tcp_flags_to_strings(flags: int) -> str:
    return "|".join(name for name, bit in TCP_FLAGS if flags & bit == bit)

def int_of_op_result(r: OpResult) -> int:
    if isinstance(r, IntResult): return r.value
    raise ValueError("Non-int op result")

def float_of_op_result(r: OpResult) -> float:
    if isinstance(r, FloatResult): return r.value
    raise ValueError("Non-float op result")

def string_of_op_result(r: OpResult) -> str:
    if isinstance(r, FloatResult): return str(r.value)
    if isinstance(r, IntResult): return str(r.value)
    if isinstance(r, IPv4Result): return str(r.value)
    if isinstance(r, MACResult): return string_of_mac(r.value)
    if isinstance(r, EmptyResult): return "Empty"
    raise ValueError("Unknown op result")

def string_of_tuple(t: TupleType) -> str:
    return ", ".join(f'"{k}" => {string_of_op_result(v)}' for k,v in t.items())

def tuple_of_list(pairs: List[TypingTuple[str, OpResult]]) -> TupleType:
    return {k:v for k,v in pairs}

# --- Lookup ---
def lookup_int(k: str, t: TupleType) -> int:
    return int_of_op_result(t[k])
def lookup_float(k: str, t: TupleType) -> float:
    return float_of_op_result(t[k])

# --- Base operators ---
init_table_size = 10000

def dump_tuple(outc=sys.stdout, show_reset: bool=False) -> Operator:
    def nxt(t: TupleType): outc.write(string_of_tuple(t)+"\n")
    def rst(t: TupleType):
        if show_reset:
            outc.write(string_of_tuple(t)+"\n"); outc.write("[reset]\n")
    return Operator(nxt, rst)

def dump_as_csv(outc=sys.stdout, header: bool=True, static_field: Optional[TypingTuple[str,str]]=None) -> Operator:
    first=True
    def nxt(t: TupleType):
        nonlocal first
        if first and header:
            if static_field: outc.write(static_field[0] + ",")
            outc.write(",".join(t.keys())+"\n"); first=False
        if static_field: outc.write(static_field[1] + ",")
        outc.write(",".join(string_of_op_result(v) for v in t.values())+"\n")
    def rst(_: TupleType): pass
    return Operator(nxt, rst)

def dump_walts_csv(filename: str) -> Operator:
    outc=None; first=True
    def nxt(t: TupleType):
        nonlocal outc, first
        if first: outc=open(filename,'w'); first=False
        row=[string_of_op_result(t[k]) for k in(
            "src_ip","dst_ip","src_l4_port","dst_l4_port",
            "packet_count","byte_count","epoch_id"
        )]
        outc.write(",".join(row)+"\n")
    def rst(_: TupleType): pass
    return Operator(nxt,rst)

def get_ip_or_zero(s: str) -> OpResult:
    return IntResult(0) if s=="0" else IPv4Result(ipaddress.IPv4Address(s))

# --- Control flow ---
def filter_op(pred: Callable[[TupleType], bool], nxt_op: Operator) -> Operator:
    def nxt(t: TupleType):
        if pred(t): nxt_op.next(t)
    def rst(t: TupleType): nxt_op.reset(t)
    return Operator(nxt, rst)

def map_op(f: Callable[[TupleType], TupleType], nxt_op: Operator) -> Operator:
    def nxt(t: TupleType): nxt_op.next(f(t))
    def rst(t: TupleType): nxt_op.reset(t)
    return Operator(nxt, rst)

def epoch(width: float, key: str, nxt_op: Operator) -> Operator:
    boundary=0.0; eid=0
    def nxt(t: TupleType):
        nonlocal boundary,eid
        tm=lookup_float("time", t)
        if boundary==0.0: boundary=tm+width
        elif tm>=boundary:
            while tm>=boundary:
                nxt_op.reset({key: IntResult(eid)}); boundary+=width; eid+=1
        t[key]=IntResult(eid); nxt_op.next(t)
    def rst(_: TupleType):
        nonlocal boundary,eid
        nxt_op.reset({key: IntResult(eid)}); boundary=0.0; eid=0
    return Operator(nxt,rst)

def meta_meter(name:str, outc=sys.stdout, static_field:Optional[str]=None, nxt_op:Operator=None) -> Operator:
    ecount=0; tcount=0
    def nxt(t: TupleType):
        nonlocal tcount; tcount+=1; nxt_op.next(t)
    def rst(t: TupleType):
        nonlocal ecount,tcount
        outc.write(f"{ecount},{name},{tcount},{static_field or ''}\n"); tcount=0; ecount+=1; nxt_op.reset(t)
    return Operator(nxt,rst)

# --- Grouping & reductions ---
def filter_groups(keys: List[str], t: TupleType) -> TupleType:
    return {k:v for k,v in t.items() if k in keys}

def single_group(_: TupleType) -> TupleType: return {}

def counter(acc: OpResult, t: TupleType) -> OpResult:
    return IntResult(1) if isinstance(acc, EmptyResult) else IntResult(acc.value+1 if isinstance(acc, IntResult) else 0)

def sum_ints(key:str, init: OpResult, t: TupleType) -> OpResult:
    total=0 if isinstance(init, EmptyResult) else init.value if isinstance(init, IntResult) else 0
    if key in t and isinstance(t[key], IntResult): total+=t[key].value; return IntResult(total)
    raise ValueError(f"sum_ints failed for {key}")

def groupby(group_fn:Callable[[TupleType],TupleType], red_fn:Callable[[OpResult,TupleType],OpResult], out_key:str, nxt_op:Operator) -> Operator:
    table:{}={}
    def nxt(t: TupleType):
        key=frozenset(group_fn(t).items()); prev=table.get(key, EmptyResult()); table[key]=red_fn(prev,t)
    def rst(base: TupleType):
        for key,val in table.items(): grp=dict(key); merged={**base,**grp, out_key: val}; nxt_op.next(merged)
        nxt_op.reset(base); table.clear()
    return Operator(nxt,rst)

def distinct(group_fn:Callable[[TupleType],TupleType], nxt_op:Operator) -> Operator:
    seen=set()
    def nxt(t: TupleType): seen.add(frozenset(group_fn(t).items()))
    def rst(base: TupleType):
        for key in seen: grp=dict(key); merged={**base,**grp}; nxt_op.next(merged)
        nxt_op.reset(base); seen.clear()
    return Operator(nxt,rst)

def split(left:Operator, right:Operator) -> Operator:
    def nxt(t: TupleType): left.next(t); right.next(t)
    def rst(t: TupleType): left.reset(t); right.reset(t)
    return Operator(nxt,rst)

# --- Join ---
def rename_filtered_keys(pairs:List[TypingTuple[str,str]], t:TupleType)->TupleType:
    return {new:t[old] for old,new in pairs if old in t}

def join(left_ex:Callable[[TupleType],TypingTuple[TupleType,TupleType]], right_ex:Callable[[TupleType],TypingTuple[TupleType,TupleType]], nxt_op:Operator, eid_key:str="eid") -> TypingTuple[Operator,Operator]:
    tbl1={}; tbl2={}; le=0; re=0
    def make_side(curr, other, ce, oe, ex):
        def nxt(t:TupleType):
            nonlocal ce, oe
            key_t, vals = ex(t); key=frozenset({**key_t, eid_key: t[eid_key]}.items())
            curr_epoch=lookup_int(eid_key,t)
            while curr_epoch>ce:
                if oe>ce: nxt_op.reset({eid_key: IntResult(ce)})
                ce+=1
            if key in other:
                ov=other.pop(key); merged={**dict(key), **vals, **ov}; nxt_op.next(merged)
            else: curr[key]=vals
        def rst(t:TupleType):
            nonlocal ce, oe
            curr_epoch=lookup_int(eid_key,t)
            while curr_epoch>ce:
                if oe>ce: nxt_op.reset({eid_key: IntResult(ce)})
                ce+=1
        return Operator(nxt,rst)
    l=make_side(tbl1,tbl2,le,re,left_ex); r=make_side(tbl2,tbl1,re,le,right_ex)
    return l,r

# --- Queries / Sonata builders ---
def ident(nxt:Operator)->Operator: return map_op(lambda t: {k:v for k,v in t.items() if k not in ("eth.src","eth.dst")}, nxt)

def count_pkts(nxt:Operator)->Operator: return epoch(1.0,"eid", groupby(single_group,counter,"pkts", nxt))

def pkts_per_src_dst(nxt:Operator)->Operator:
    return epoch(1.0,"eid", groupby(lambda t: filter_groups(["ipv4.src","ipv4.dst"],t), counter, "pkts", nxt))

def distinct_srcs(nxt:Operator)->Operator:
    return epoch(1.0,"eid", distinct(lambda t: filter_groups(["ipv4.src"],t), groupby(single_group,counter,"srcs", nxt)))

def tcp_new_cons(nxt:Operator)->Operator:
    return epoch(1.0,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==2, groupby(lambda tt: filter_groups(["ipv4.dst"],tt), counter,"cons", nxt)))

def ssh_brute_force(nxt:Operator)->Operator:
    return epoch(1.0,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.dport",t)==22, distinct(lambda tt: filter_groups(["ipv4.src","ipv4.dst","ipv4.len"],tt), groupby(lambda tt: filter_groups(["ipv4.dst","ipv4.len"],tt), counter,"srcs", nxt))))

def super_spreader(nxt:Operator)->Operator:
    return epoch(1.0,"eid", distinct(lambda t: filter_groups(["ipv4.src","ipv4.dst"],t), groupby(lambda t: filter_groups(["ipv4.src"],t), counter,"dsts", nxt)))

def port_scan(nxt:Operator)->Operator:
    return epoch(1.0,"eid", distinct(lambda t: filter_groups(["ipv4.src","l4.dport"],t), groupby(lambda t: filter_groups(["ipv4.src"],t), counter,"ports", nxt)))

def ddos(nxt:Operator)->Operator:
    return epoch(1.0,"eid", distinct(lambda t: filter_groups(["ipv4.src","ipv4.dst"],t), groupby(lambda t: filter_groups(["ipv4.dst"],t), counter,"srcs", nxt)))

def syn_flood_sonata(nxt:Operator)->List[Operator]:
    th=3; dur=1.0
    # syns, synacks, acks
    syns=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==2, groupby(lambda tt: filter_groups(["ipv4.dst"],tt), counter,"syns", n)))
    synacks=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==18, groupby(lambda tt: filter_groups(["ipv4.src"],tt), counter,"synacks", n)))
    acks=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==16, groupby(lambda tt: filter_groups(["ipv4.dst"],tt), counter,"acks", n)))
    # first join
    j1,j2 = join(lambda t: (filter_groups(["host"],t), filter_groups(["syns+synacks"],t)),
                  lambda t: (rename_filtered_keys([("ipv4.dst","host")],t), filter_groups(["acks"],t)),
                  map_op(lambda t: {**t, "syns+synacks-acks": IntResult(lookup_int("syns+synacks",t)-lookup_int("acks",t))}, nxt))
    # second join
    j3,j4= join(lambda t: (rename_filtered_keys([("ipv4.dst","host")],t), filter_groups(["syns"],t)),
                lambda t: (rename_filtered_keys([("ipv4.src","host")],t), filter_groups(["synacks"],t)),
                j1)
    return [syns(j3), synacks(j4), acks(j2)]

def completed_flows(nxt:Operator)->List[Operator]:
    th=1; dur=30.0
    syn=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==2, groupby(lambda tt: filter_groups(["ipv4.dst"],tt), counter,"syns", n)))
    fin=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and (lookup_int("l4.flags",t)&1)==1, groupby(lambda tt: filter_groups(["ipv4.src"],tt), counter,"fins", n)))
    o1,o2= join(lambda t: (rename_filtered_keys([("ipv4.dst","host")],t), filter_groups(["syns"],t)),
                 lambda t: (rename_filtered_keys([("ipv4.src","host")],t), filter_groups(["fins"],t)),
                 map_op(lambda t: {**t, "diff": IntResult(lookup_int("syns",t)-lookup_int("fins",t))}, nxt))
    return [syn(o1), fin(o2)]

def slowloris(nxt:Operator)->List[Operator]:
    t1,t2,t3=5,500,90; dur=1.0
    ncon=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6, distinct(lambda tt: filter_groups(["ipv4.src","ipv4.dst","l4.sport"],tt), groupby(lambda tt: filter_groups(["ipv4.dst"],tt), counter,"n_conns", n))))
    nbytes=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6, groupby(lambda tt: filter_groups(["ipv4.dst"],tt), sum_ints("ipv4.len", EmptyResult(), tt),"n_bytes", n)))
    o1,o2= join(lambda t:(filter_groups(["ipv4.dst"],t), filter_groups(["n_conns"],t)),
                lambda t:(filter_groups(["ipv4.dst"],t), filter_groups(["n_bytes"],t)),
                map_op(lambda t: {**t, "bytes_per_conn": IntResult(lookup_int("n_bytes",t)//lookup_int("n_conns",t))}, nxt))
    return [ncon(o1), nbytes(o2)]

def join_test(nxt:Operator)->List[Operator]:
    dur=1.0
    syn=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==2, n))
    synacks=lambda n: epoch(dur,"eid", filter_op(lambda t: lookup_int("ipv4.proto",t)==6 and lookup_int("l4.flags",t)==18, n))
    o1,o2= join(lambda t:(rename_filtered_keys([("ipv4.src","host")],t), rename_filtered_keys([("ipv4.dst","remote")],t)),
                lambda t:(rename_filtered_keys([("ipv4.dst","host")],t), filter_groups(["time"],t)), nxt)
    return [syn(o1), synacks(o2)]

def q3(nxt:Operator)->Operator: return epoch(100.0,"eid", distinct(lambda t: filter_groups(["ipv4.src","ipv4.dst"],t), nxt))

def q4(nxt:Operator)->Operator: return epoch(10000.0,"eid", groupby(lambda t: filter_groups(["ipv4.dst"],t), counter,"pkts", nxt))

# --- Assemble and run queries ---
queries: List[Operator] = [ident(dump_tuple())]

def run_queries() -> None:
    for i in range(20):
        tup = {
            "time": FloatResult(float(i)),
            "eth.src": MACResult(b"\x00\x11\x22\x33\x44\x55"),
            "eth.dst": MACResult(b"\xAA\xBB\xCC\xDD\xEE\xFF"),
            "eth.ethertype": IntResult(0x0800),
            "ipv4.hlen": IntResult(20),
            "ipv4.proto": IntResult(6),
            "ipv4.len": IntResult(60),
            "ipv4.src": IPv4Result(ipaddress.IPv4Address("127.0.0.1")),
            "ipv4.dst": IPv4Result(ipaddress.IPv4Address("127.0.0.1")),
            "l4.sport": IntResult(440),
            "l4.dport": IntResult(50000),
            "l4.flags": IntResult(10)
        }
        for op in queries:
            op.next(tup)
    print("Done")

if __name__ == "__main__": run_queries()

File: LLM-linguistic-translations/X-grok/3-think/ts-translations/chat-code.ts
// Additional Utility Functions
function filterGroups(inclKeys: string[], tup: Tuple): Tuple {
    const newTup = new Map<string, OpResult>();
    inclKeys.forEach(key => {
      const value = tup.get(key);
      if (value) newTup.set(key, value);
    });
    return newTup;
  }
  
  function sumInts(searchKey: string, initVal: OpResult, tup: Tuple): OpResult {
    if (initVal.kind === 'Empty') return { kind: 'Int', value: 0 };
    if (initVal.kind === 'Int') {
      const value = tup.get(searchKey);
      if (value && value.kind === 'Int') {
        return { kind: 'Int', value: initVal.value + value.value };
      }
      throw new Error(`'sumInts' failed to find integer value for "${searchKey}"`);
    }
    return initVal;
  }
  
  // Additional Operator Creators
  function split(l: Operator, r: Operator): Operator {
    return {
      next: (tup: Tuple) => { l.next(tup); r.next(tup); },
      reset: (tup: Tuple) => { l.reset(tup); r.reset(tup); },
    };
  }
  
  type KeyExtractor = (tup: Tuple) => [Tuple, Tuple];
  
  function join(
    leftExtractor: KeyExtractor,
    rightExtractor: KeyExtractor,
    nextOp: Operator
  ): [Operator, Operator] {
    const hTbl1 = new Map<string, Tuple>();
    const hTbl2 = new Map<string, Tuple>();
    let leftCurrEpoch = 0;
    let rightCurrEpoch = 0;
  
    function handleJoinSide(
      currHTbl: Map<string, Tuple>,
      otherHTbl: Map<string, Tuple>,
      currEpochRef: { value: number },
      otherEpochRef: { value: number },
      f: KeyExtractor
    ): Operator {
      return {
        next: (tup: Tuple) => {
          const [key, vals] = f(tup);
          const currEpoch = lookupInt('eid', tup);
          while (currEpoch > currEpochRef.value) {
            if (otherEpochRef.value > currEpochRef.value) {
              nextOp.reset(new Map([['eid', { kind: 'Int', value: currEpochRef.value }]]));
            }
            currEpochRef.value++;
          }
          const newTup = new Map([...key, ['eid', { kind: 'Int', value: currEpoch }]]);
          const keyStr = serializeTuple(newTup);
          const otherVal = otherHTbl.get(keyStr);
          if (otherVal) {
            otherHTbl.delete(keyStr);
            const unionedTup = new Map([...newTup, ...vals, ...otherVal]);
            nextOp.next(unionedTup);
          } else {
            currHTbl.set(keyStr, vals);
          }
        },
        reset: (tup: Tuple) => {
          const currEpoch = lookupInt('eid', tup);
          while (currEpoch > currEpochRef.value) {
            if (otherEpochRef.value > currEpochRef.value) {
              nextOp.reset(new Map([['eid', { kind: 'Int', value: currEpochRef.value }]]));
            }
            currEpochRef.value++;
          }
        },
      };
    }
  
    const leftOp = handleJoinSide(hTbl1, hTbl2, { value: leftCurrEpoch }, { value: rightCurrEpoch }, leftExtractor);
    const rightOp = handleJoinSide(hTbl2, hTbl1, { value: rightCurrEpoch }, { value: leftCurrEpoch }, rightExtractor);
    return [leftOp, rightOp];
  }
  
  function renameFilteredKeys(renamings: [string, string][], inTup: Tuple): Tuple {
    const newTup = new Map<string, OpResult>();
    renamings.forEach(([oldKey, newKey]) => {
      const value = inTup.get(oldKey);
      if (value) newTup.set(newKey, value);
    });
    return newTup;
  }
  
  // Query Definitions
  function pktsPerSrcDst(nextOp: Operator): Operator {
    return epoch(1.0, 'eid', groupby(
      (tup: Tuple) => filterGroups(['ipv4.src', 'ipv4.dst'], tup),
      counter,
      'pkts',
      nextOp
    ));
  }
  
  function distinctSrcs(nextOp: Operator): Operator {
    return epoch(1.0, 'eid', distinct(
      (tup: Tuple) => filterGroups(['ipv4.src'], tup),
      groupby(singleGroup, counter, 'srcs', nextOp)
    ));
  }
  
  function tcpNewCons(nextOp: Operator): Operator {
    const threshold = 40;
    return epoch(1.0, 'eid', filter(
      (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 2,
      groupby(
        (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
        counter,
        'cons',
        filter(
          (tup: Tuple) => lookupInt('cons', tup) >= threshold,
          nextOp
        )
      )
    ));
  }
  
  function sshBruteForce(nextOp: Operator): Operator {
    const threshold = 40;
    return epoch(1.0, 'eid', filter(
      (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.dport', tup) === 22,
      distinct(
        (tup: Tuple) => filterGroups(['ipv4.src', 'ipv4.dst', 'ipv4.len'], tup),
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.dst', 'ipv4.len'], tup),
          counter,
          'srcs',
          filter(
            (tup: Tuple) => lookupInt('srcs', tup) >= threshold,
            nextOp
          )
        )
      )
    ));
  }
  
  function superSpreader(nextOp: Operator): Operator {
    const threshold = 40;
    return epoch(1.0, 'eid', distinct(
      (tup: Tuple) => filterGroups(['ipv4.src', 'ipv4.dst'], tup),
      groupby(
        (tup: Tuple) => filterGroups(['ipv4.src'], tup),
        counter,
        'dsts',
        filter(
          (tup: Tuple) => lookupInt('dsts', tup) >= threshold,
          nextOp
        )
      )
    ));
  }
  
  function portScan(nextOp: Operator): Operator {
    const threshold = 40;
    return epoch(1.0, 'eid', distinct(
      (tup: Tuple) => filterGroups(['ipv4.src', 'l4.dport'], tup),
      groupby(
        (tup: Tuple) => filterGroups(['ipv4.src'], tup),
        counter,
        'ports',
        filter(
          (tup: Tuple) => lookupInt('ports', tup) >= threshold,
          nextOp
        )
      )
    ));
  }
  
  function ddos(nextOp: Operator): Operator {
    const threshold = 45;
    return epoch(1.0, 'eid', distinct(
      (tup: Tuple) => filterGroups(['ipv4.src', 'ipv4.dst'], tup),
      groupby(
        (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
        counter,
        'srcs',
        filter(
          (tup: Tuple) => lookupInt('srcs', tup) >= threshold,
          nextOp
        )
      )
    ));
  }
  
  function synFloodSonata(nextOp: Operator): Operator[] {
    const threshold = 3;
    const epochDur = 1.0;
  
    const syns = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 2,
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
          counter,
          'syns',
          nextOp
        )
      ));
  
    const synacks = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 18,
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.src'], tup),
          counter,
          'synacks',
          nextOp
        )
      ));
  
    const acks = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 16,
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
          counter,
          'acks',
          nextOp
        )
      ));
  
    const [joinOp1, joinOp2] = join(
      (tup: Tuple) => [
        filterGroups(['host'], tup),
        filterGroups(['syns+synacks'], tup)
      ],
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.dst', 'host']], tup),
        filterGroups(['acks'], tup)
      ],
      map(
        (tup: Tuple) => {
          const newTup = new Map(tup);
          newTup.set('syns+synacks-acks', {
            kind: 'Int',
            value: lookupInt('syns+synacks', tup) - lookupInt('acks', tup)
          });
          return newTup;
        },
        filter(
          (tup: Tuple) => lookupInt('syns+synacks-acks', tup) >= threshold,
          nextOp
        )
      )
    );
  
    const [joinOp3, joinOp4] = join(
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.dst', 'host']], tup),
        filterGroups(['syns'], tup)
      ],
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.src', 'host']], tup),
        filterGroups(['synacks'], tup)
      ],
      map(
        (tup: Tuple) => {
          const newTup = new Map(tup);
          newTup.set('syns+synacks', {
            kind: 'Int',
            value: lookupInt('syns', tup) + lookupInt('synacks', tup)
          });
          return newTup;
        },
        joinOp1
      )
    );
  
    return [
      syns(joinOp3),
      synacks(joinOp4),
      acks(joinOp2)
    ];
  }
  
  function completedFlows(nextOp: Operator): Operator[] {
    const threshold = 1;
    const epochDur = 30.0;
  
    const syns = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 2,
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
          counter,
          'syns',
          nextOp
        )
      ));
  
    const fins = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && (lookupInt('l4.flags', tup) & 1) === 1,
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.src'], tup),
          counter,
          'fins',
          nextOp
        )
      ));
  
    const [op1, op2] = join(
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.dst', 'host']], tup),
        filterGroups(['syns'], tup)
      ],
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.src', 'host']], tup),
        filterGroups(['fins'], tup)
      ],
      map(
        (tup: Tuple) => {
          const newTup = new Map(tup);
          newTup.set('diff', {
            kind: 'Int',
            value: lookupInt('syns', tup) - lookupInt('fins', tup)
          });
          return newTup;
        },
        filter(
          (tup: Tuple) => lookupInt('diff', tup) >= threshold,
          nextOp
        )
      )
    );
  
    return [syns(op1), fins(op2)];
  }
  
  function slowloris(nextOp: Operator): Operator[] {
    const t1 = 5;
    const t2 = 500;
    const t3 = 90;
    const epochDur = 1.0;
  
    const nConns = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6,
        distinct(
          (tup: Tuple) => filterGroups(['ipv4.src', 'ipv4.dst', 'l4.sport'], tup),
          groupby(
            (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
            counter,
            'n_conns',
            filter(
              (tup: Tuple) => lookupInt('n_conns', tup) >= t1,
              nextOp
            )
          )
        )
      ));
  
    const nBytes = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6,
        groupby(
          (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
          (acc, tup) => sumInts('ipv4.len', acc, tup),
          'n_bytes',
          filter(
            (tup: Tuple) => lookupInt('n_bytes', tup) >= t2,
            nextOp
          )
        )
      ));
  
    const [op1, op2] = join(
      (tup: Tuple) => [
        filterGroups(['ipv4.dst'], tup),
        filterGroups(['n_conns'], tup)
      ],
      (tup: Tuple) => [
        filterGroups(['ipv4.dst'], tup),
        filterGroups(['n_bytes'], tup)
      ],
      map(
        (tup: Tuple) => {
          const newTup = new Map(tup);
          newTup.set('bytes_per_conn', {
            kind: 'Int',
            value: Math.floor(lookupInt('n_bytes', tup) / lookupInt('n_conns', tup))
          });
          return newTup;
        },
        filter(
          (tup: Tuple) => lookupInt('bytes_per_conn', tup) <= t3,
          nextOp
        )
      )
    );
  
    return [nConns(op1), nBytes(op2)];
  }
  
  function joinTest(nextOp: Operator): Operator[] {
    const epochDur = 1.0;
  
    const syns = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 2,
        nextOp
      ));
  
    const synacks = (nextOp: Operator) =>
      epoch(epochDur, 'eid', filter(
        (tup: Tuple) => lookupInt('ipv4.proto', tup) === 6 && lookupInt('l4.flags', tup) === 18,
        nextOp
      ));
  
    const [op1, op2] = join(
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.src', 'host']], tup),
        renameFilteredKeys([['ipv4.dst', 'remote']], tup)
      ],
      (tup: Tuple) => [
        renameFilteredKeys([['ipv4.dst', 'host']], tup),
        filterGroups(['time'], tup)
      ],
      nextOp
    );
  
    return [syns(op1), synacks(op2)];
  }
  
  function q3(nextOp: Operator): Operator {
    return epoch(100.0, 'eid', distinct(
      (tup: Tuple) => filterGroups(['ipv4.src', 'ipv4.dst'], tup),
      nextOp
    ));
  }
  
  function q4(nextOp: Operator): Operator {
    return epoch(10000.0, 'eid', groupby(
      (tup: Tuple) => filterGroups(['ipv4.dst'], tup),
      counter,
      'pkts',
      nextOp
    ));
  }
  
  // Execute Queries
  function runQueries(): void {
    const queries: Operator[] = [ident(dumpTupleOperator())];
    const tuples = Array.from({ length: 20 }, (_, i) => {
      const tup = new Map<string, OpResult>();
      tup.set('time', { kind: 'Float', value: 0.000000 + i });
      tup.set('eth.src', { kind: 'MAC', value: '\x00\x11\x22\x33\x44\x55' });
      tup.set('eth.dst', { kind: 'MAC', value: '\xAA\xBB\xCC\xDD\xEE\xFF' });
      tup.set('eth.ethertype', { kind: 'Int', value: 0x0800 });
      tup.set('ipv4.hlen', { kind: 'Int', value: 20 });
      tup.set('ipv4.proto', { kind: 'Int', value: 6 });
      tup.set('ipv4.len', { kind: 'Int', value: 60 });
      tup.set('ipv4.src', { kind: 'IPv4', value: '127.0.0.1' });
      tup.set('ipv4.dst', { kind: 'IPv4', value: '127.0.0.1' });
      tup.set('l4.sport', { kind: 'Int', value: 440 });
      tup.set('l4.dport', { kind: 'Int', value: 50000 });
      tup.set('l4.flags', { kind: 'Int', value: 10 });
      return tup;
    });
    tuples.forEach(tup => queries.forEach(query => query.next(tup)));
    console.log('Done');
  }
  
  // Execute
  runQueries();